File: Java/src/ch08/IsBalance.java
package ch08;

public class IsBalance {
    
    public static class Node{
        public int value;
        public Node left;
        public Node right;

        public Node(int v){
            value = v;
        }
    }
    
    public static class Info{
        public boolean isBalance;
        public int height;
        
        public Info(boolean b,int h){
            isBalance = b;
            height = h;
        }
    }
    
    public static Info process(Node root){
        if (null == root){
            return new Info(true,0);
        }
        
        Info leftInfo = process(root.left);
        Info rightInfo = process(root.right);
        
        int height = Math.max(leftInfo.height,rightInfo.height) + 1;
        boolean isBalance = false;
        if (leftInfo.isBalance && rightInfo.isBalance && (Math.abs(leftInfo.height - rightInfo.height) <= 1)){
            isBalance = true;
        }
        return new Info(isBalance,height);
    }


    public static void main(String[] args) {
        Node root = new Node(6);
        root.left = new Node(4);
        root.right = new Node(8);
        root.left.left = new Node(2);
        root.left.right = new Node(5);
        root.right.left = new Node(7);
        root.right.right = new Node(9);
        root.left.left.left = new Node(1);
        root.left.left.right = new Node(3);

        System.out.println(process(root).isBalance);
    }
    
}

File: Python/ch08/IsBalance.py
class Node:
    def __init__(self, v):
        self.value = v
        self.left = None
        self.right = None


class Info:
    def __init__(self, b, h):
        self.is_balance = b
        self.height = h


def process(r):
    if r is None:
        return Info(True, 0)

    left_info = process(r.left)
    right_info = process(r.right)

    height = max(left_info.height, right_info.height) + 1
    is_balance = False
    if left_info.is_balance and right_info.is_balance and abs(left_info.height - right_info.height) <= 1:
        is_balance = True
    return Info(is_balance, height)


if __name__ == '__main__':
    root = Node(6)
    root.left = Node(4)
    root.right = Node(8)
    root.left.left = Node(2)
    root.left.right = Node(5)
    root.right.left = Node(7)
    root.right.right = Node(9)
    root.left.left.left = Node(1)
    root.left.left.right = Node(3)

    print(process(root).is_balance)
